pkg/storegateway/indexheader/reader_pool.go

```diff
@@ -49,11 +53,54 @@ type ReaderPool struct {
 	// Keep track of all readers managed by the pool.
 	lazyReadersMx sync.Mutex
 	lazyReaders   map[*LazyBinaryReader]struct{}
+	tracker       HeadersLazyLoadedTracker
+}
+
+// HeadersLazyLoadedTrackerState tracks blocks that are lazy loaded and its last access time.
+type HeadersLazyLoadedTrackerState struct {
+	sync.Mutex
+	state storepb.IndexHeaderTrackerState
+}
+
+// HeadersLazyLoadedTracker persists the list of lazy loaded blocks.
+type HeadersLazyLoadedTracker struct {
+	lazyReaders map[*LazyBinaryReader]struct{}
+	userID      string
+	persistPath string
+}
+
+// Write writes the list of lazy loaded block to the writer.
+func (h *HeadersLazyLoadedTracker) Write() error {
+	trackerState := HeadersLazyLoadedTrackerState{
+		state: storepb.IndexHeaderTrackerState{
+			LazyLoadedBlocks: make(map[string]uint32),
+			UserId:           h.userID,
+		},
+	}
+
+	trackerState.Lock()
+	defer trackerState.Unlock()
+
+	for k := range h.lazyReaders {
```

Comment:
If we want to access `h.lazyReaders` we should be holding the `ReaderPool.lazyReadersMx`. 

IDK how you feel, but I'm a bit against sharing maps and a mutex in a struct. One idea is to change the interface of `HeadersLazyLoadedTracker` to accept a `map[string]int64` and take care of checksumming and persisting it (later of loading it too)?

Reply:
I forgot about lazyReaderMx. Will call the lock before reading it.

For the mutex in the tracker, woouldn't we are going to remove it?